syncapi/internal/keychange.go

```diff
@@ -215,15 +216,15 @@ func TrackChangedUsers(
 	return changed, left, nil
 }
 
+// filterSharedUsers takes a list of remote users whose keys have changed and filters
+// it down to include only users who the requesting user shares a room with.
 func filterSharedUsers(
-	ctx context.Context, rsAPI roomserverAPI.SyncRoomserverAPI, userID string, usersWithChangedKeys []string,
+	ctx context.Context, db storage.SharedUsers, userID string, usersWithChangedKeys []string,
 ) (map[string]int, []string) {
-	var result []string
-	var sharedUsersRes roomserverAPI.QuerySharedUsersResponse
-	err := rsAPI.QuerySharedUsers(ctx, &roomserverAPI.QuerySharedUsersRequest{
-		UserID:       userID,
-		OtherUserIDs: usersWithChangedKeys,
-	}, &sharedUsersRes)
+	sharedUsersRes := &roomserverAPI.QuerySharedUsersResponse{
+		UserIDsToCount: map[string]int{},
+	}
+	sharedUsers, err := db.SharedUsers(ctx, userID, usersWithChangedKeys)
```

Comment:
These users aren't in the map above anymore. Is this okay?

Reply:
I think you've highlighted a bug in the map handling there which I've now fixed.